src/BodyPipe.cc

```diff
@@ -266,18 +266,18 @@ BodyPipe::clearConsumer()
 void
 BodyPipe::expectNoConsumption()
 {
-    // We may be called multiple times because multiple jobs on the consumption
-    // chain may realize that there will be no more setConsumer() calls (e.g.,
-    // consuming code and retrying code). It is both difficult and not really
-    // necessary for them to coordinate their expectNoConsumption() calls.
-
-    // As a consequence, we may be called when we are auto-consuming already.
-
-    if (!abortedConsumption && !exhausted()) {
-        // Before we abort, any regular consumption should be over and auto
-        // consumption must not be started.
-        Must(!theConsumer);
-
+    // We and enableAutoConsumption() may be called multiple times because
+    // multiple jobs on the consumption chain may realize that there will be no
+    // more setConsumer() calls (e.g., consuming code and retrying code). It is
+    // both difficult and unnecessary for them to coordinate their calls.
+
+    // As a consequence, we may be called when already auto-consuming, including
+    // cases where abortedConsumption is still false. We could try to harden
+    // this by also aborting consumption from enableAutoConsumption() when there
+    // is no consumer, but see errorAppendEntry() TODO for a better plan.
+
+    debugs(91, 7, status());
+    if (!abortedConsumption && !exhausted() && !theConsumer) {
```

Comment:
Why is `theConsumer` a consideration here? surely auto-consume by this `BodyPipe` instance should not matter whether one exists still?

Reply:
> Why is `theConsumer` a consideration here?

This `if` statement tells the producer that pipe consumption has aborted (among other things). A pipe should not tell the producer that consumption has aborted when there is a consumer. If, instead of moving theConsumer precondition from an assertion to `if` statement, we drop that precondition, then the producer will receive this incorrect/unwanted noteBodyConsumerAborted notification via asyncCall() below this line.

There are two primary cases here:

* expectNoConsumption() caller calls when there is no auto-consumption is going on. We are fine in this case: theConsumer will be nil, and the `if` statement will notify the producer that there will be no consumption.
* expectNoConsumption() caller calls when auto-consumption is already going on (but the caller is not aware of that fact). Official code asserts. PR code does nothing. This subtle [terminology] conflict/mess is not ideal, but this is still OK because the producer does not need to be notified that there will be no consumption since there _is_ consumption.

> surely auto-consume by this BodyPipe instance should not matter whether one exists still?

I am not sure I understand this part of the question, but please note that BodyPipe auto-consumption is implemented using a special consumer job. In other words, if "auto-consumption is currently active (i.e. going on)", then theConsumer is set. This PR adds a TODO that, if successfully implemented, will simplify/clarify this messy situation quite a bit.

Reply:
Thank you for explaining.